aggregator_core/src/datastore.rs

```diff
@@ -5321,6 +5401,130 @@ pub mod models {
+        Batch<SEED_SIZE, Q, A>
+    {
+        /// Creates a new [`Batch`].
+        pub fn new(
+            task_id: TaskId,
+            batch_identifier: Q::BatchIdentifier,
+            aggregation_parameter: A::AggregationParam,
+            state: BatchState,
+            outstanding_aggregation_jobs: u64,
+        ) -> Self {
+            Self {
+                task_id,
+                batch_identifier,
+                aggregation_parameter,
+                state,
+                outstanding_aggregation_jobs,
+            }
+        }
+
+        /// Gets the task ID associated with this batch.
+        pub fn task_id(&self) -> &TaskId {
+            &self.task_id
+        }
+
+        /// Gets the batch identifier associated with this batch.
+        pub fn batch_identifier(&self) -> &Q::BatchIdentifier {
+            &self.batch_identifier
+        }
+
+        /// Gets the aggregation parameter associated with this batch.
+        pub fn aggregation_parameter(&self) -> &A::AggregationParam {
+            &self.aggregation_parameter
+        }
+
+        /// Gets the state associated with this batch.
+        pub fn state(&self) -> &BatchState {
+            &self.state
+        }
+
+        /// Returns a new batch equivalent to the current batch, but with the given state.
+        pub fn with_state(self, state: BatchState) -> Self {
+            Self { state, ..self }
+        }
+
+        /// Gets the count of outstanding aggregation jobs associated with this batch.
+        pub fn outstanding_aggregation_jobs(&self) -> u64 {
+            self.outstanding_aggregation_jobs
+        }
+
+        /// Returns a new batch equivalent to the current batch, but with the given count of
+        /// outstanding aggregation jobs.
+        pub fn with_outstanding_aggregation_jobs(self, outstanding_aggregation_jobs: u64) -> Self {
+            Self {
+                outstanding_aggregation_jobs,
+                ..self
+            }
+        }
+    }
+
+    impl<const SEED_SIZE: usize, Q: QueryType, A: vdaf::Aggregator<SEED_SIZE, 16>> PartialEq
+        for Batch<SEED_SIZE, Q, A>
+    where
+        Q::BatchIdentifier: PartialEq,
+        A::AggregationParam: PartialEq,
+    {
+        fn eq(&self, other: &Self) -> bool {
+            self.task_id == other.task_id
+                && self.batch_identifier == other.batch_identifier
+                && self.aggregation_parameter == other.aggregation_parameter
+                && self.state == other.state
+                && self.outstanding_aggregation_jobs == other.outstanding_aggregation_jobs
+        }
+    }
+
+    impl<const SEED_SIZE: usize, Q: QueryType, A: vdaf::Aggregator<SEED_SIZE, 16>> Eq
+        for Batch<SEED_SIZE, Q, A>
+    where
+        Q::BatchIdentifier: Eq,
+        A::AggregationParam: Eq,
+    {
+    }
+
+    impl<const SEED_SIZE: usize, Q: QueryType, A: vdaf::Aggregator<SEED_SIZE, 16>> Hash
+        for Batch<SEED_SIZE, Q, A>
```

Comment:
humorously(?), fixing _this_ manual `Hash` impl caused `rustc` to panic until I cleaned cache (`cargo` & `docker` both). this did not occur fixing the other `Hash`. 🤷🏻

Reply:
oh, this doesn't work at all, actually: it seems that to acceptably derive `Hash`, `Hash` would also need to be implemented for the VDAF type in use (e.g. the `Prio3` or `Poplar1` types from `libprio-rs`). I guess that makes sense, in general: two structs with differing type parameters will generally be considered inequal.

`libprio-rs` could be modified to guarantee that VDAF types provide `Hash`, but there are a whole bunch of other traits that we should probably include, too (basically, everything we require in our aggregator methods' huge `where` blocks). But I think last time we asked, we got pushback because these traits aren't inherently required for a VDAF implementation.

IIRC, the `[Partial]Eq` situation was similar.

I reverted the hash change to use the manual hash.